language/move-stdlib/src/natives/debug.rs

```diff
@@ -60,29 +61,61 @@ fn native_print(
     #[cfg(feature = "testing")]
     {
         let ty = ty_args.pop().unwrap();
-
-        match is_string_type(_context, &ty, move_std_addr) {
-            true => {
-                let r = pop_arg!(args, StructRef);
-
-                let field_ref = r
-                    .borrow_field(0)? // borrows the 'bytes' field of the std::string::String struct
-                    .value_as::<VectorRef>()?;
-
-                let bytes = field_ref.as_bytes_ref();
-
-                // This is safe because we guarantee the bytes to be utf8.
-                let str = unsafe { std::str::from_utf8_unchecked(bytes.as_slice()) };
-
-                println!("[debug] {}", str);
-            }
-            false => {
-                let r = pop_arg!(args, Reference);
-                let mut buf = String::new();
-                print_reference(&mut buf, &r)?;
-                println!("[debug] {}", buf);
+        
+        // if ty is a struct type:
+        // get type layout in VM format (raw BCS, no field names) 
+        let old_type_layout = context.type_to_type_layout(&ty)?.unwrap();
+
+        // println!("Type: {:?}", &ty);
+        // println!("Value: {:?}", &args[0]);
+        // println!("MoveTypeLayout: {:?}", old_type_layout);
+
+        match &old_type_layout {
+            MoveTypeLayout::Struct(old_struct_layout) => {
+                let arg = pop_arg!(args, StructRef);
+                let arg_val = arg.read_ref()?;
+
+                let blob = arg_val.simple_serialize(&old_type_layout).unwrap();
+
+                // get BCS bytes for the struct
+                let move_struct = MoveStruct::simple_deserialize(&blob, old_struct_layout).unwrap();
+
+                // get type layout in decorated format (with field names)
+                let new_type_layout = context.type_to_fully_annotated_layout(&ty)?.unwrap();
+                let new_struct_layout = match &new_type_layout {
+                    MoveTypeLayout::Struct(struct_layout) => struct_layout,
+                    _ => panic!("ooppsie; this shouldn't happen.")
+                };
+
+                let decorated_struct = move_struct.decorate(new_struct_layout);
+                println!("[annotated-debug] {}", decorated_struct);
+            },
+            _ => {
+                match is_string_type(context, &ty, move_std_addr) {
```

Comment:
First time I've been looking through this code, so sorry if its a duplicate question, but, shouldn't we just add some custom printing/display logic for this? If we don't, it means we will print strings as strings when they are top level values, but not deeply inside of structs

Reply:
Yep, agreed with this. We will also not print fields of nested structs unless we go in this direction.

Reply:
Indeed! I should've been more clear that this is a draft PR. So all of it will change!